Separate loading logic from configuration

tomwayson opened the pull request:
This is a first attempt to move reusable loader logic out of load.ts and into new library esri-system-js/esriLoader.js. The idea is to make it easier for people to use this pattern in their applications. Ideally, I would say we should go so far as to move esri-system-js out of this repo into it's own repo/package and make adding it as easy as `npm install esri-system-js`.

But first, I wanted to validate the idea w/ you here, and if you're into, settle on the names, etc. For now I've extracted your TS code as JS exposed as a global on `window`. If we do end up extracting it into its own repo, it could be written in TS, ES6, etc and distributed as a UMD module? Not sure what makes the most sense (i.e. if there are other non-Angular/TS use cases).

I'm open to suggestions on any of this. The main thing is provide a way for users to get your magic System.js sauce w/o having to clone this whole repo and copy files, update config etc.

Thoughts? Feelings?

cc @jwasilgeo

tomwayson committed:
make loading of esri modules configurable

move reusable loader logic out of load.ts and into
new library esri-system-js/esriLoader.js, which:
- separates loading logic from configuration (which remains in load.ts)
- tries to reduce the global vars it creates to just one (esriLoader)
added comments to index.html and load.ts
for now, esriLoader is in JS instead of TS, so updated gitignore to
include it

jwasilgeo commented:
@odoe, first of all, super cool work here with the System.register paradigm.  On my own, I had enough of a painful hurdle fighting with making System configs and Dojo's require play nice, and thankfully have been able to re-purpose the code you cracked in **load.ts** for other Angular2 needs.

Perhaps not fully my place to say, but throwing in a couple thoughts: @tomwayson this commit is pretty neat.  As for the naming collision [possibility](https://github.com/tomwayson/esrijs4-vm-angular2/blob/esri-system-js/esri-system-js/esriLoader.js#L5), could another optional argument be added to `esriLoader.require` that's an array (in the same order) of the desired output module names?  Maybe not a perfect idea, but something to think about.

tomwayson commented:
@jwasilgeo I had a similar thought to handle the potential name collisions. I was thinking you could pass in an optional hash w/ as part of the config that contains only the names for modules that you want to override the default. For example, to make 'esri/request' come out 'esriRequest' you would:

``` js
esriLoader.require({
  moduleName: 'esri-mods',
  moduleNameOverrides: {
    'esri/request': 'esriRequest'  
  }
}, [
  'esri/Map',
  'esri/views/MapView',
  'esri/widgets/Home/HomeViewModel',
  'esri/request'
], () => {
  // then bootstrap application
  System.import('app/boot')
    .then(null, console.error.bind(console));  
});
```

Also, a couple other thoughts I've had on the signature/name. First, I don't like how the options are the first argument unless that becomes an optional config argument (much like dojo/require). [That's my plan](https://github.com/tomwayson/esrijs4-vm-angular2/blob/esri-system-js/esri-system-js/esriLoader.js#L28), just hadn't implemented it yet. But I'm also open to suggestion here - like moving it to the end and making it optional.

If this is only ever going to do this one thing, it could probably just expose a function (rather than a namespace obj w/ a function) that could be called `esriRequire()` or `esriRegister()` or `registerEsriModules()`. Again, open to suggestion on any of that.

odoe commented:
@tomwayson @jwasilgeo 
I like the idea of providing the alias names in the options. I think that would make it easier to avoid any name collisions. Better than letting the regex extract the module name or concatenating the path to a name, which would get unwieldily quickly.

This looks really cool!

tomwayson committed:
replaced local esri-system-js w/ npm install

moved esri-system-js into it's own npm package and
installed it as a dependency
deleted local files, updated href in index.html, and
updated function call in load.ts to new signature

tomwayson commented:
@odoe 

I've moved esri-system-js into it's own repo/[package](https://www.npmjs.com/package/esri-system-js) and published. now this PR pulls it in from there.

tomwayson committed:
updated esri-system-js ver

jwasilgeo commented:
Classic PR thread, brings back memories!
